### ISSUE TITLE
yolov8-pose multi-class detection problem

### ISSUE THREAD

**Issue by brian2023707** [ultralytics/ultralytics#17489]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, may I ask if YOLOv8-Pose can be trained for multi-class keypoint detection? My dataset consists of images with both hands in the frame, and I would like to detect the keypoints for the left and right hands separately.

I am currently using Labelme to annotate by creating rectangles for `right_hand` and `left_hand`, setting the group ID as 0 and 1, and assigning group IDs for the keypoints of each hand. Then I convert the annotations to YOLOv8-Pose format like this:

```
0 <x> <y> <width> <height> <px1> <py1> <px2> <py2> ... <pxn> <pyn>
1 <x> <y> <width> <height> <px1> <py1> <px2> <py2> ... <pxn> <pyn>
```

However, during training, an error occurs indicating that the number of label lines is incorrect.

If YOLOv8-Pose does not support multi-class keypoint detection, what should I do to achieve separate keypoint detection for the left and right hands?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @brian2023707, thank you for reaching out and showing interest in Ultralytics 🚀! We appreciate your inquiry about YOLOv8-Pose and its capabilities.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug the issue. Having a sample code snippet or dataset could greatly assist in diagnosing the problem.

If this is a custom training ❓ Question, kindly provide as much detail as possible, including dataset samples, the specific error messages you're encountering, and any training logs. Also, ensure that you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to potentially improve your setup.

We recommend checking out the [Docs](https://docs.ultralytics.com) for comprehensive guides and examples, especially if you're new to the platform. You can find relevant usage examples for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) there.

For real-time assistance, consider joining our community on [Discord](https://ultralytics.com/discord) 🎧. If you prefer more detailed discussions or to share your insights, visit our [Discourse](https://community.ultralytics.com) or our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Make sure you're using the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) within a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to check if your issue persists in the newest version:

```bash
pip install -U ultralytics
```

## Environments

Feel free to test your setup in any of the following supported environments, complete with all dependencies like [CUDA](https://developer.nvidia.com/cuda), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/):

- **Notebooks** with free GPU access: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: Refer to [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: Refer to [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: Refer to [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

When this badge is green, it means all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. These tests ensure the smooth operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across various platforms like macOS, Windows, and Ubuntu daily and on every code commit.

Rest assured, an Ultralytics engineer will review your issue soon and assist you further! 😊

**Comment by Y-T-G** [comment]
Yes, you can have multiple classes. But all classes should have same number of keypoints.

**Comment by brian2023707** [comment]
May I ask if my YAML configuration is correct? I've labeled a bounding box and seven keypoints for each of the left and right hands.

```
kpt_shape: [14, 2]
flip_idx: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]

names:
  0: right_hand
  1: left_hand
```

Based on the above configuration, I received the feedback:
'yolo8 pose train ignoring corrupt image/label: labels require 33 columns each.'

I think the issue is that the model expects only one class, so the calculation would be 1 + 4 + 14*2 = 33."

**Comment by brian2023707** [comment]
My label looks like this:
```
0 <x> <y> <width> <height> <px1> <py1> <px2> <py2> ... <pxn> <pyn>
1 <x> <y> <width> <height> <px1> <py1> <px2> <py2> ... <pxn> <pyn>
```
Since I have two classes, and each class has seven keypoints along with a bounding box, there should be 19 columns and 2 rows.

**Comment by Y-T-G** [comment]
If it has 7 keypoints, then why are you specifying the kpt_shape to be [14,2]. That implies you have 14 keypoints for each class. It should be [7,2]. Change it and delete the labels.cache files in the train and val folders and run again.

**Comment by Y-T-G** [comment]
Also did you actually check the label files like I said? It can be 19 columns, but it can also be 1 + 4 + 7*3 = 26 columns depending on whether the keypoints also include visibility labels. Check the labels to see what the actual number is.
